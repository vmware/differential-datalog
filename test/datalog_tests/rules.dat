start;
insert R3(0, true),
insert R3(1, true),
insert R3(2, true),
insert R3(3, true),
insert R3(4, true),
insert R3(5, true),
insert R3(6, true);
commit;
dump;

start;
insert Table1(0, "foo"),
insert Table1(1, "foo"),
insert Table1(0xabcdef000, "foo"),
insert Table1(125, "buzzzzzzzz");
insert Table2("foo", 1234567890, S{(true,true),5}),
insert Table2("foo", 12, S{(true,false),0xabcd}),
insert Table2("foo", 100000000000000, S{(false,true),10000}),
insert Table2("buzzzzzzzz", 1234567890, S{(true,true),5});
commit;
dump Table12;
