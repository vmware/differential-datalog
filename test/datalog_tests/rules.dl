typedef R1 = bigint

input relation R1[R1]
input relation R2[(bigint, bigint)]

R1[1].
R2[(1,2)].

input relation R3(f1: bigint, f2: bool)
relation R4[R3]
relation R5[bigint]
relation R6(f: bigint)

R4[x] :- R3[x].
R5[x.f1] :- R3[x].
R6(x.f1) :- R3[x].

input relation X(b: bigint, c: R3, d: R1)
relation Z(x: bigint, y: R3)
Z(a.b, a.c) :- X[a].

input relation Y(b: bigint, e: string)
relation T(x: bigint, y: string)

T(a.b, e) :- X[a], Y(a.d, e).

typedef R7 = R7{f1: bigint, f2: bit<16>}
           | R7Other

relation R7[R7]

R7(1,2).
R7[R7{1,3}].
R7[R7Other].

typedef S = S {
    f1: (bool, bool),
    f2: bit<32>
}
input relation Table1(id: bit<64>, name: string)
input relation Table2(name: string, f2: bigint, f3: S)

relation Table12(id: bit<128>, name: string, f3: S)

Table12(hash128((id,name,f2,f3)), name, f3) :- Table1(id, name), Table2(name, f2, f3).
