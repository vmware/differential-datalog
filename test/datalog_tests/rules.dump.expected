R3:
R3{0,true}
R3{1,true}
R3{2,true}
R3{3,true}
R3{4,true}
R3{5,true}
R3{6,true}

R4:
R3{0,true}
R3{1,true}
R3{2,true}
R3{3,true}
R3{4,true}
R3{5,true}
R3{6,true}

R5:
0
1
2
3
4
5
6

R6:
R6{0}
R6{1}
R6{2}
R6{3}
R6{4}
R6{5}
R6{6}

Table12{20695836920908937261915266253713563738,"foo",S{(true, true),5}}
Table12{81692126345696542188923640651411714412,"foo",S{(false, true),10000}}
Table12{137151609739491206170748566711757625861,"foo",S{(true, false),43981}}
Table12{173592105249357153805573863620167017533,"foo",S{(false, true),10000}}
Table12{173670172741911619502105830072953828992,"foo",S{(true, false),43981}}
Table12{211727904196777888792921590151204575764,"foo",S{(true, true),5}}
Table12{218334107078988268330177571792522447425,"foo",S{(true, true),5}}
Table12{258101594679877739861006822360175003366,"buzzzzzzzz",S{(true, true),5}}
Table12{312560568469280405643575197132576203606,"foo",S{(false, true),10000}}
Table12{324905179504175058909924909437678210347,"foo",S{(true, false),43981}}
